compiler/rustc_typeck/src/coherence/orphan.rs

```diff
@@ -275,59 +275,84 @@ fn emit_orphan_check_error<'tcx>(
             err.note("define and implement a trait or new type instead");
             err.emit()
         }
-        traits::OrphanCheckErr::UncoveredTy(param_ty, local_type) => {
-            let mut sp = sp;
-            for param in generics.params {
-                if param.name.ident().to_string() == param_ty.to_string() {
-                    sp = param.span;
+        traits::OrphanCheckErr::UncoveredTy(param_ty, idx, local_type) => {
+            let sp;
+            let mut param_ty_in_fundamental_ty = false;
+
+            match idx {
+                0 => {
+                    sp = self_ty_span;
+                    match *self_ty.kind() {
+                        ty::Ref(..) => {
+                            param_ty_in_fundamental_ty = true;
+                        }
+                        ty::Adt(def, _) => {
+                            param_ty_in_fundamental_ty = def.is_fundamental();
+                        }
+                        _ => (),
+                    }
                 }
+                _ => sp = generics[idx - 1].span(),
             }
 
+            let message = match *param_ty.kind() {
+                ty::Projection(..) => "associated type",
+                _ => "type parameter",
+            };
+
+            let fundamental_ty_msg =
+                if param_ty_in_fundamental_ty { "as argument to a fundamental type " } else { "" };
+
             match local_type {
                 Some(local_type) => struct_span_err!(
                     tcx.sess,
                     sp,
                     E0210,
-                    "type parameter `{}` must be covered by another type \
+                    "{} `{}` {}must be covered by another type \
```

Comment:
this is existing, but these messages would probably be a lot clearer if we used inline format args, i.e.

> "{message} `{param_ty}` {fundamental_ty_msg}must be covered by another type \

(renaming `message` to `param_ty_kind` or something like that)